### ISSUE TITLE
./ Is used as a prefix in all specs

### ISSUE THREAD

**Issue by ishubin** [galenframework/galen#318]

**Comment by adamjson** [comment]
Does this ticket track the issue discussed here: https://groups.google.com/forum/#!topic/galen-framework/ulw5D_Dd8UA ?

We are unable to include and run JavaScript, ultimately due to a FileNotFoundException. It looks like the absolute file path is being prepended with ".\". )(See stacktrace below).

I was able to track the problem down to JsFunctionLoad.load(...). My temporary local workaround was to comment out the else statement so that line 91 (fullPath = contextPath + File.separator + filePath) is never executed. I'm not sure that this is an OS / use case agnostic fix since my ramp up time to get the project building, testing, and executing locally has been slow. 

This issue has been a blocker for us, and I'd like to help work on this ticket but my time has been very limited lately.

Any insight is appreciated.

Cheers.

Stack:

java.lang.RuntimeException: Could not load script: .\C:\Users\<user_omitted>\Documents\Repos\<path_omitted>\main-content/test.js
        at com.galenframework.javascript.JsFunctionLoad.load(JsFunctionLoad.java:118)
        at com.galenframework.javascript.GalenJsExecutor.runJavaScriptFromFile(GalenJsExecutor.java:165)
        at com.galenframework.speclang2.pagespec.PageSpecHandler.runJavaScriptFromFile(PageSpecHandler.java:432)
        at com.galenframework.speclang2.pagespec.ScriptProcessor.process(ScriptProcessor.java:50)
        at com.galenframework.speclang2.pagespec.MacroProcessor.processMacroStatement(MacroProcessor.java:200)
        at com.galenframework.speclang2.pagespec.MacroProcessor.process(MacroProcessor.java:80)
        at com.galenframework.speclang2.pagespec.PageSpecReader.read(PageSpecReader.java:64)
        at com.galenframework.speclang2.pagespec.PageSpecReader.read(PageSpecReader.java:47)
        at com.galenframework.api.Galen.checkLayout(Galen.java:68)
        at com.galenframework.api.Galen.checkLayout(Galen.java:59)
        at com.galenframework.api.Galen.checkLayout(Galen.java:154)
        at com.galenframework.api.Galen.checkLayout(Galen.java:128)
        at <...stacktrace_cut...>
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
        at java.lang.reflect.Method.invoke(Unknown Source)
        at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
        at org.testng.internal.Invoker.invokeMethod(Invoker.java:639)
        at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:816)
        at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1124)
        at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
        at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)
Caused by: java.io.FileNotFoundException: .\C:\Users\<user_omitted>\Documents\Repos\<path_omitted>\main-content/test.js
        at com.galenframework.javascript.JsFunctionLoad.load(JsFunctionLoad.java:106)

**Comment by ishubin** [comment]
@adamjson I think I missed this one. I am planning to fix it for current version. I already have some minor changed waiting to be released, so I guess I will just fix this story and release everything into v 2.2.2.

**Comment by ishubin** [comment]
@hypery2k I pushed a fix for this ./ context path thingy. Could you please check it on your Windows machine? I want to make sure we got rid of it completely

@adamjson could you please also help testing this one?

**Comment by adamjson** [comment]
Sure, I'll take a look this afternoon sometime.

**Comment by hypery2k** [comment]
I will check it tomorrow

**Comment by hypery2k** [comment]
relative paths are working so far, but absolute paths, like C:\ ... not:

```
com.galenframework.parser.FileSyntaxException: Cannot find file: \specs\shared/C:\galen-extras-rules.gspec
```

I think we should add some unit tests like in my branch, see my comment on your commit. I think we should add an if check around, like this one: https://github.com/galenframework/galen/commit/8c3a4e328bf9ac4d788d167bd85a29458df8ca92#diff-cbf0f1db27794662f7251ae56e1a3979R116

**Comment by ishubin** [comment]
@hypery2k I don't understand this to be honest. How can I reproduce that last error? How could it happen  that happen anyway that C: gets into that string? Could you please share the code that reproduces this issue?

**Comment by hypery2k** [comment]
If I use something like this in the Spec:
```
@import C:\galen-extras-rules.gspec
```

**Comment by ishubin** [comment]
Oh, but that is a wrong usage of galen specs. We don't have support of importing absolute paths. All imports are performed relatively to the current location of original spec file.
As for Js executor - we allow to load path starting with `/` but that is also not absolute path. That tells galen to start file search in the root of the project (working directory)

**Comment by hypery2k** [comment]
ok, got it ;)

**Comment by ishubin** [comment]
Actually in future I would like to refactor all js and spec loaders to always use "/" and be completely abstracted from what operating system they are running on. But that is a bit complicated for now.

So is this issue fixed? Do you encounter any other similar issues on Windows machine? Because then I can release an update today

**Comment by hypery2k** [comment]
no, so far it's working on Windows 7 and Windows 10 for me

**Comment by ishubin** [comment]
@hypery2k Ok, then I will wait for @adamjson and, if everything is ok, I will release an update. In a meantime I will play with appveyor and will fix the tests, so that we can launch them in Windows

**Comment by adamjson** [comment]
I'll be setting up the test in a few miunutes. I got sucked into a problem with TestNG yesterday and ran out if time.

**Comment by adamjson** [comment]
@ishubin The fix looks good from my end, and I haven't seen any regressions (comparing results from our last run with 2.2.3 against 2.2.5-SNAPSHOT).

**Comment by ishubin** [comment]
@adamjson Thanks a lot! I will try to release a version 2.2.5 today then.

**Comment by ishubin** [comment]
I have released version 2.2.5 with this fix. Closing the issue
